test/acceptance/js/FilestoreApp.js

```diff
@@ -1,109 +1,112 @@
- * DS101: Remove unnecessary use of Array.from
- * DS102: Remove unnecessary code created because of implicit returns
- * DS103: Rewrite code to no longer use __guard__
- * DS205: Consider reworking code to avoid use of IIFEs
- * DS207: Consider shorter variations of null checks
- * Full docs: https://github.com/decaffeinate/decaffeinate/blob/master/docs/suggestions.md
- */
-const app = require('../../../app')
-require('logger-sharelatex').logger.level('info')
 const logger = require('logger-sharelatex')
 const Settings = require('settings-sharelatex')
+const fs = require('fs')
+const Path = require('path')
 const request = require('request')
+const { promisify } = require('util')
+const disrequire = require('disrequire')
 
 const S3_TRIES = 30
 
-module.exports = {
-  running: false,
-  initing: false,
-  callbacks: [],
-  ensureRunning(callback) {
-    if (callback == null) {
-      callback = function(error) {}
-    }
+logger.logger.level('info')
+
+const fsReaddir = promisify(fs.readdir)
+
+class FilestoreApp {
+  constructor() {
+    this.running = false
+    this.initing = false
+  }
+
+  async runServer() {
     if (this.running) {
-      return callback()
-    } else if (this.initing) {
-      return this.callbacks.push(callback)
-    } else {
-      this.initing = true
-      this.callbacks.push(callback)
-      return app.listen(
-        __guard__(
-          Settings.internal != null ? Settings.internal.filestore : undefined,
-          x => x.port
-        ),
+      return
+    }
+
+    if (this.initing) {
+      return this.waitForInit()
+    }
+    this.initing = true
+
+    this.app = await FilestoreApp.requireApp()
+
+    await new Promise((resolve, reject) => {
+      this.server = this.app.listen(
+        Settings.internal.filestore.port,
         'localhost',
-        error => {
-          if (error != null) {
-            throw error
+        err => {
+          if (err) {
+            return reject(err)
           }
-          this.running = true
-          logger.log('filestore running in dev mode')
-
-          return (() => {
-            const result = []
-            for (callback of Array.from(this.callbacks)) {
-              result.push(callback())
-            }
-            return result
-          })()
+          resolve()
         }
       )
+    })
+
+    if (Settings.filestore.backend === 's3') {
+      try {
+        await FilestoreApp.waitForS3()
+      } catch (err) {
+        await this.stop()
+        throw err
+      }
+    }
+
+    this.initing = false
+  }
+
+  async waitForInit() {
+    while (this.initing) {
+      await new Promise(resolve => setTimeout(resolve, 1000))
     }
-  },
-
-  waitForS3(callback, tries) {
-    if (
-      !(Settings.filestore.s3 != null
-        ? Settings.filestore.s3.endpoint
-        : undefined)
-    ) {
-      return callback()
+  }
+
+  async stop() {
+    if (this.server) {
+      await new Promise(resolve => {
+        this.server.close(resolve)
```

Comment:
There's no error handling if something goes wrong closing the server. How about:
```
const closeServer = promisify(this.server.close).bind(this.server)
try {
  await closeServer()
} finally {
  delete this.server
}
```

Reply:
nice :+1: